File: 20-Mediator/src/go/chat_user.go
package main

import "fmt"

// ChatUser is a Concrete Colleague. It holds only the mediator, never a
// reference to another user.
type ChatUser struct {
	name     string
	mediator ChatMediator
}

func NewChatUser(mediator ChatMediator, name string) *ChatUser {
	return &ChatUser{name: name, mediator: mediator}
}

func (u *ChatUser) Name() string { return u.name }

// Send routes the message through the mediator, not to peers directly.
func (u *ChatUser) Send(message string) {
	fmt.Printf("%s sends: %s\n", u.name, message)
	u.mediator.SendMessage(message, u)
}

func (u *ChatUser) Receive(message, from string) {
	fmt.Printf("    %s received from %s: %s\n", u.name, from, message)
}

File: 20-Mediator/src/java/chapter20/ChatUser.java
package chapter20;

/**
 * Concrete Colleague — a chat participant. It only implements how it reacts
 * to an incoming message; it has no reference to any other user.
 */
public class ChatUser extends User {
    public ChatUser(ChatMediator mediator, String name) {
        super(mediator, name);
    }

    @Override
    public void receive(String message, String from) {
        System.out.println("    " + name + " received from " + from + ": " + message);
    }
}
